benchmarks/cpp/tensorexpr/bench_reduce.cpp

```diff
@@ -379,3 +383,220 @@ BENCHMARK_DEFINE_F(Reduce1D, TeRfactorV1)(benchmark::State& state) {
+    cg.call({A.data_ptr<float>(), B.data_ptr<float>()});
+  }
+}
+BENCHMARK_REGISTER_F(Reduce2DCol, OpSchedule)->Apply(//CustomArgs);
+    [](benchmark::internal::Benchmark* b) {
+      for (auto sch : {0, 1, 2, 3}) {
+        for (auto rows : {3, 6, 12}) {
+          auto cols = 24 - rows;
+          b->Args({1 << rows, 1 << cols, sch});
+        }
+      }
+    });
+
+class Reduce2DRow : public benchmark::Fixture {
+ public:
+  void SetUp(const benchmark::State& state) override {
+    at::set_num_threads(1);
+    torch::manual_seed(0x12345678);
+    M = state.range(0);
+    N = state.range(1);
+    A = torch::randn({M, N});
+    ref = torch::sum(A, {1});
+    B = torch::zeros_like(ref);
+  }
+
+  void TearDown(benchmark::State& state) override {
+    TORCH_CHECK(at::allclose(B, ref, std::sqrt(A.numel()) * 1e-4));
+    state.counters["BYTES"] = benchmark::Counter(uint64_t(state.iterations()) * (A.nbytes() + B.nbytes()),
+                                                 benchmark::Counter::kIsRate);
+  }
+
+  int M;
+  int N;
+  at::Tensor A;
+  at::Tensor B;
+  at::Tensor ref;
+};
+
+BENCHMARK_DEFINE_F(Reduce2DRow, Torch)(benchmark::State& state) {
+  for (auto _ : state) {
+    B = torch::sum(A, {1});
+  }
+}
+BENCHMARK_REGISTER_F(Reduce2DRow, Torch)
+->Args({1 << 3, 1 << 21})
+->Args({1 << 6, 1 << 18})
+->Args({1 << 12, 1 << 12})
+->Args({1 << 18, 1 << 6});
+
+BENCHMARK_DEFINE_F(Reduce2DRow, Hand)(benchmark::State& state) {
+  auto a = A.data_ptr<float>();
+  auto b = B.data_ptr<float>();
+  constexpr int Mb = 4;
+  constexpr int Nb = 4;
+  auto fn = [&] {
+    for (int m_outer = 0; m_outer < M; m_outer += Mb) {
+      float bregs[Mb][Nb] = {0.0f};
+      for (int n_outer = 0; n_outer < N; n_outer += Nb) {
+        for (int m_inner = 0; m_inner < Mb; m_inner++) {
+          for (int n_inner = 0; n_inner < Nb; n_inner++) {
+            bregs[m_inner][n_inner] += a[(m_outer + m_inner) * N + n_outer + n_inner];
+          }
+        }
+      }
+      for (int m_inner = 0; m_inner < Mb; m_inner++) {
+        b[m_outer + m_inner] = 0.f;
+        for (int n_inner = 0; n_inner < Nb; n_inner++) {
+          b[m_outer + m_inner] += bregs[m_inner][n_inner];
+        }
+      }
+    }
+  };
+  for (auto _ : state) {
+    fn();
+  }
+}
+BENCHMARK_REGISTER_F(Reduce2DRow, Hand)
+->Args({1 << 18, 1 << 6});
+
+BENCHMARK_DEFINE_F(Reduce2DRow, OpSchedule)(benchmark::State& state) {
+  te::KernelScope ks;
+  constexpr int kChunkSize = 8;
+  te::Placeholder a("A", te::kFloat, {M, N});
+  te::Tensor* b = te::computeSum({a.handle(), te::IntList({1}), false}, at::kFloat);
+  te::LoopNest nest({b});
+
+  auto sch = state.range(2);
+  if (sch == 1) {
+    auto loops = nest.getLoopStmtsFor(b);
+    te::For *mi, *mo;
+    te::Buf *rf;
+    nest.splitWithMask(loops[1], kChunkSize, &mi);
+    loops = nest.reorder({loops[1], mi}, {1, 0});
+    TORCH_CHECK(nest.rfactor(nest.getLoopBodyFor(b), loops[0], &rf));
+    nest.reorderAxis(loops[0], loops[1]);
```

Comment:
May not be relevant here, but still FYI. `reorderAxis` does not perform in-place transformation, so the loop pointers will not be valid after the transformation. On the other hand, `reorder` performs the transformation in-place, in case you want to reuse the loop pointers afterwards.

However, if the loopnest between the loops that are being reordered is not perfectly nested, then `reorderAxis` is the only option. `reorder` expects the loopnest to be perfectly nested.